src/operator/nn/batch_norm.cu

```diff
@@ -708,7 +708,7 @@ void BatchNormGradCompute<gpu>(const nnvm::NodeAttrs& attrs,
     })
   }
 #else
-  MSHADOW_REAL_TYPE_SWITCH_EX(out_grad[0].type_flag_, DType, AccReal, {
+  MSHADOW_REAL_TYPE_SWITCH_EX(dtype, DType, AccReal, {
```

Comment:
@tdomhan please have a look here

Reply:
I guess the dtype of the inputs and outputs must always be the same. Using the dtype variable (as done in this PR) seems cleaner.